controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,173 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package twophase
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/utils"
+)
+
+type chaosStateMachine struct {
+	Chaos v1alpha1.InnerSchedulerObject
+	Req   ctrl.Request
+	*Reconciler
+}
+
+func unexpectedInto(currentPhase v1alpha1.ExperimentPhase, targetPhase v1alpha1.ExperimentPhase) error {
+	return errors.Errorf("turn from %s into %s is unexpected", currentPhase, targetPhase)
+}
+
+// This method changes the phase of an object and do some side effects
+// There are 6 different phases, so there could be 6 * 6 = 36 branches
+func (m *chaosStateMachine) run(ctx context.Context, targetPhase v1alpha1.ExperimentPhase) (bool, error) {
+	updated := false
+
+	duration, err := m.Chaos.GetDuration()
+	if err != nil {
+		m.Log.Error(err, "failed to get chaos duration")
+		return updated, err
+	}
+	if duration == nil {
+		zero := time.Duration(0)
+		duration = &zero
+	}
+	if m.Chaos.GetScheduler() == nil {
+		m.Log.Info("Scheduler should be defined currently")
+		return updated, fmt.Errorf("misdefined scheduler")
+	}
+
+	status := m.Chaos.GetStatus()
+	currentPhase := m.Chaos.GetStatus().Experiment.Phase
+	m.Log.Info("change phase", "current phase", currentPhase, "target phase", targetPhase)
+
+	// covers 6 conditions
+	if currentPhase == targetPhase {
+		return updated, nil
+	}
+
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseFinished {
+		return updated, unexpectedInto(currentPhase, targetPhase)
+	}
+
+	// In this branch `targetPhase != v1alpha1.ExperimentPhaseRunning` is always true
+	// move from Running phase to any phase should recover the experiment.
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseRunning {
+		updated = true
+
+		m.Log.Info("recovering", "current phase", currentPhase, "target phase", targetPhase)
+		err := m.Recover(ctx, m.Req, m.Chaos)
+		if err != nil {
+			status.FailedMessage = err.Error()
+
+			m.Log.Error(err, "fail to recover")
+			return updated, err
+		}
+		status.Experiment.Phase = targetPhase
+
+		now := time.Now()
+		status.Experiment.EndTime = &metav1.Time{
+			Time: now,
+		}
+		status.Experiment.Duration = now.Sub(status.Experiment.StartTime.Time).String()
+
+		// Pause from a running phase should set the nextStart to now
+		// so that the Reconciler will start another cycle of running right
+		// after resume
```

Comment:
if the `cron` is like `30 * * * *`, it means run chaos on the 30th minute every hour,  can't run right after resume

Reply:
@cwen0 I don't know the exact definition of `pause`. It seems that the `pause` has two different meanings for these two kinds of scheduler.

Maybe we should clarify it first, through an RFC or an issue... I could modify the implementation according to that later (in another PR). Confusing feature will lead to confusing codes, which we always hate.

Reply:
I don't think `pause` has two meanings for two kinds of schedulers. For the current implementation, in both situations, if we resume at the waiting state of the scheduler, we would turn the state from paused to waiting (at "Recovering"), get the next start time from the scheduler  ("Waiting"), wait for the next start time and go into "Starting". 
So it's like, the scheduler keeps the same, and the pause period is dug out of it.

For #1207, I checked [the original implementation](https://github.com/chaos-mesh/chaos-mesh/pull/15/commits/dd87ad8600ebac800fae8f54fc43aa3ef24ed78d). Maybe I missed something, but I think change nextRecover from `now + duration` to `nextStart + duration` could fix it.